File: python/137_Single_Number_II.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        res = 0
        
        for i in range(32):
            mask = 1 << i
            cnt = 0
            for j in range(len(nums)):
                if (nums[j] & mask) != 0:
                    cnt += 1
            if cnt % 3 != 0:
                # why python needs to do extra judgement here?
                if i == 31:
                    res -= mask
                else:
                    res |= mask
        
        return res

File: java/137_Single_Number_II.java
class Solution {
    public int singleNumber(int[] nums) {
        return singleNumber2(nums);
    }

    public int singleNumber1(int[] nums) {
        int res = 0;
        for(int i = 0 ; i < 32 ; i++){
            int mask = 1<< i;
            int cnt = 0;
            for(int j = 0 ; j < nums.length; j++){
                if((nums[j] & mask) != 0){
                    cnt++;
                }
            }
            if (cnt % 3 != 0){
                res |= mask;
            }
        }
        return res;
    }

    public int singleNumber2(int[] nums) {
        int a = 0, b = 0;
        for ( int num : nums){
            b = (b ^ num) & ~a;
            a = (a ^ num) & ~b;
        }
        return b;
    }
}